File: RailLog/RailLog/Services/BaiduTrainTicketOcrService.cs
using System.Globalization;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;
using RailLog.Models;

namespace RailLog.Services;

public sealed class BaiduTrainTicketOcrService(IHttpClientFactory httpClientFactory, IConfiguration configuration)
{
    public const int MaxImageBytes = 5 * 1024 * 1024;

    private const string TokenEndpoint = "oauth/2.0/token";
    private const string TicketOcrEndpoint = "rest/2.0/ocr/v1/train_ticket";

    public async Task<TrainTicketOcrResult> RecognizeAsync(byte[] imageBytes, CancellationToken cancellationToken = default)
    {
        if (imageBytes.Length == 0)
        {
            return TrainTicketOcrResult.Failed("上传图片不能为空。");
        }

        if (imageBytes.Length > MaxImageBytes)
        {
            return TrainTicketOcrResult.Failed($"图片大小不能超过 {MaxImageBytes / 1024 / 1024}MB。");
        }

        var apiKey = configuration["BaiduOcr:ApiKey"];
        var secretKey = configuration["BaiduOcr:SecretKey"];
        if (string.IsNullOrWhiteSpace(apiKey) || string.IsNullOrWhiteSpace(secretKey))
        {
            return TrainTicketOcrResult.Failed("未配置百度 OCR Key。请在配置中设置 BaiduOcr:ApiKey 和 BaiduOcr:SecretKey。");
        }

        var client = httpClientFactory.CreateClient("BaiduOcrApi");
        var tokenResult = await GetAccessTokenAsync(client, apiKey, secretKey, cancellationToken);
        if (!tokenResult.Success || string.IsNullOrWhiteSpace(tokenResult.AccessToken))
        {
            return TrainTicketOcrResult.Failed(tokenResult.ErrorMessage ?? "获取 OCR Access Token 失败。");
        }

        var imageBase64 = Convert.ToBase64String(imageBytes);
        using var request = new HttpRequestMessage(HttpMethod.Post, $"{TicketOcrEndpoint}?access_token={Uri.EscapeDataString(tokenResult.AccessToken)}")
        {
            Content = new FormUrlEncodedContent(new Dictionary<string, string>
            {
                ["image"] = imageBase64
            })
        };
        request.Headers.Accept.ParseAdd("application/json");

        using var response = await client.SendAsync(request, cancellationToken);
        var raw = await response.Content.ReadAsStringAsync(cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            return TrainTicketOcrResult.Failed($"OCR 请求失败（HTTP {(int)response.StatusCode}）。");
        }

        BaiduTrainTicketOcrApiResponse? parsed;
        try
        {
            parsed = JsonSerializer.Deserialize<BaiduTrainTicketOcrApiResponse>(raw);
        }
        catch (JsonException)
        {
            return TrainTicketOcrResult.Failed("OCR 响应解析失败。");
        }

        if (parsed is null)
        {
            return TrainTicketOcrResult.Failed("OCR 响应为空。");
        }

        if (parsed.ErrorCode.HasValue)
        {
            return TrainTicketOcrResult.Failed($"OCR 服务错误：{parsed.ErrorMessage ?? parsed.ErrorCode.Value.ToString(CultureInfo.InvariantCulture)}");
        }

        if (parsed.WordsResult is null)
        {
            return TrainTicketOcrResult.Failed("未识别到车票字段。");
        }

        var words = parsed.WordsResult;
        return TrainTicketOcrResult.Succeeded(new TrainTicketOcrDto
        {
            TrainNumber = NormalizeTrainNumber(words.TrainNumber),
            TravelDate = ParseDate(words.Date),
            DepartureTime = ParseTime(words.Time),
            FromStation = NormalizeStation(words.StartingStation),
            ToStation = NormalizeStation(words.DestinationStation),
            SeatType = NormalizeText(words.SeatCategory),
            SeatNumber = NormalizeText(words.SeatNumber),
            Price = ParsePrice(words.TicketRates)
        });
    }

    private static async Task<TokenResult> GetAccessTokenAsync(HttpClient client, string apiKey, string secretKey, CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, TokenEndpoint)
        {
            Content = new FormUrlEncodedContent(new Dictionary<string, string>
            {
                ["grant_type"] = "client_credentials",
                ["client_id"] = apiKey,
                ["client_secret"] = secretKey
            })
        };
        using var response = await client.SendAsync(request, cancellationToken);
        var raw = await response.Content.ReadAsStringAsync(cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            return TokenResult.Failed($"获取 Access Token 失败（HTTP {(int)response.StatusCode}）。");
        }

        BaiduTokenResponse? parsed;
        try
        {
            parsed = JsonSerializer.Deserialize<BaiduTokenResponse>(raw);
        }
        catch (JsonException)
        {
            return TokenResult.Failed("Access Token 响应解析失败。");
        }

        if (parsed is null || string.IsNullOrWhiteSpace(parsed.AccessToken))
        {
            return TokenResult.Failed(parsed?.ErrorDescription ?? "Access Token 为空。");
        }

        return TokenResult.Succeeded(parsed.AccessToken);
    }

    private static string NormalizeText(string? value)
    {
        return value?.Trim() ?? string.Empty;
    }

    private static string NormalizeTrainNumber(string? value)
    {
        var text = NormalizeText(value).Replace(" ", string.Empty, StringComparison.Ordinal).ToUpperInvariant();
        if (string.IsNullOrWhiteSpace(text))
        {
            return string.Empty;
        }

        var match = Regex.Match(text, @"[A-Z]\d{1,4}");
        if (match.Success)
        {
            return match.Value;
        }

        return text;
    }

    private static string NormalizeStation(string? value)
    {
        var text = NormalizeText(value);
        if (text.Length > 1 && text.EndsWith("站", StringComparison.Ordinal))
        {
            return text[..^1];
        }

        return text;
    }

    private static DateOnly? ParseDate(string? value)
    {
        var text = NormalizeText(value);
        if (string.IsNullOrWhiteSpace(text))
        {
            return null;
        }

        var formats = new[] { "yyyy年MM月dd日", "yyyy-M-d", "yyyy-MM-dd", "yyyy/M/d", "yyyy/MM/dd" };
        if (DateOnly.TryParseExact(text, formats, CultureInfo.InvariantCulture, DateTimeStyles.None, out var date))
        {
            return date;
        }

        return null;
    }

    private static TimeOnly? ParseTime(string? value)
    {
        var text = NormalizeText(value);
        if (string.IsNullOrWhiteSpace(text))
        {
            return null;
        }

        var formats = new[] { "HH:mm", "H:mm" };
        if (TimeOnly.TryParseExact(text, formats, CultureInfo.InvariantCulture, DateTimeStyles.None, out var time))
        {
            return time;
        }

        return null;
    }

    private static decimal? ParsePrice(string? value)
    {
        var text = NormalizeText(value);
        if (string.IsNullOrWhiteSpace(text))
        {
            return null;
        }

        var match = Regex.Match(text, @"\d+(?:\.\d+)?");
        if (!match.Success)
        {
            return null;
        }

        if (!decimal.TryParse(match.Value, NumberStyles.Number, CultureInfo.InvariantCulture, out var price))
        {
            return null;
        }

        return price > 0 ? price : null;
    }

    private sealed record TokenResult(bool Success, string? AccessToken, string? ErrorMessage)
    {
        public static TokenResult Succeeded(string token) => new(true, token, null);
        public static TokenResult Failed(string message) => new(false, null, message);
    }

    private sealed class BaiduTokenResponse
    {
        [JsonPropertyName("access_token")]
        public string? AccessToken { get; set; }

        [JsonPropertyName("error_description")]
        public string? ErrorDescription { get; set; }
    }

    private sealed class BaiduTrainTicketOcrApiResponse
    {
        [JsonPropertyName("words_result")]
        public BaiduTrainTicketWordsResult? WordsResult { get; set; }

        [JsonPropertyName("error_code")]
        public int? ErrorCode { get; set; }

        [JsonPropertyName("error_msg")]
        public string? ErrorMessage { get; set; }
    }

    private sealed class BaiduTrainTicketWordsResult
    {
        [JsonPropertyName("train_num")]
        public string? TrainNumber { get; set; }

        [JsonPropertyName("date")]
        public string? Date { get; set; }

        [JsonPropertyName("time")]
        public string? Time { get; set; }

        [JsonPropertyName("starting_station")]
        public string? StartingStation { get; set; }

        [JsonPropertyName("destination_station")]
        public string? DestinationStation { get; set; }

        [JsonPropertyName("seat_category")]
        public string? SeatCategory { get; set; }

        [JsonPropertyName("seat_num")]
        public string? SeatNumber { get; set; }

        [JsonPropertyName("ticket_rates")]
        public string? TicketRates { get; set; }
    }
}

File: app/lib/src/services/baidu_train_ticket_ocr_service.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:dio/dio.dart';
import 'package:shared_preferences/shared_preferences.dart';

class BaiduOcrCredentials {
  const BaiduOcrCredentials({required this.apiKey, required this.secretKey});

  final String apiKey;
  final String secretKey;

  bool get isComplete =>
      apiKey.trim().isNotEmpty && secretKey.trim().isNotEmpty;
}

abstract final class BaiduOcrSettings {
  static const _apiKeyPreference = 'baidu_ocr_api_key';
  static const _secretKeyPreference = 'baidu_ocr_secret_key';

  static Future<BaiduOcrCredentials> load() async {
    final preferences = await SharedPreferences.getInstance();
    return BaiduOcrCredentials(
      apiKey: preferences.getString(_apiKeyPreference) ?? '',
      secretKey: preferences.getString(_secretKeyPreference) ?? '',
    );
  }

  static Future<void> save({
    required String apiKey,
    required String secretKey,
  }) async {
    final preferences = await SharedPreferences.getInstance();
    await preferences.setString(_apiKeyPreference, apiKey.trim());
    await preferences.setString(_secretKeyPreference, secretKey.trim());
  }
}

class TrainTicketOcrResult {
  const TrainTicketOcrResult({
    required this.trainNumber,
    required this.fromStation,
    required this.toStation,
    required this.departureTime,
    required this.seatType,
    required this.seatNumber,
    required this.price,
  });

  final String trainNumber;
  final String fromStation;
  final String toStation;
  final DateTime departureTime;
  final String? seatType;
  final String? seatNumber;
  final double? price;
}

class BaiduTrainTicketOcrService {
  BaiduTrainTicketOcrService({Dio? dio}) : _dio = dio ?? Dio();

  final Dio _dio;

  Future<TrainTicketOcrResult> recognize(
    Uint8List imageBytes,
    BaiduOcrCredentials credentials,
  ) async {
    if (!credentials.isComplete) {
      throw const BaiduOcrException('请先在设置中填写百度 OCR API Key 和 Secret Key');
    }
    final tokenResponse = await _dio.post<Map<String, dynamic>>(
      'https://aip.baidubce.com/oauth/2.0/token',
      queryParameters: {
        'grant_type': 'client_credentials',
        'client_id': credentials.apiKey.trim(),
        'client_secret': credentials.secretKey.trim(),
      },
    );
    final accessToken = tokenResponse.data?['access_token']?.toString();
    if (accessToken == null || accessToken.isEmpty) {
      throw BaiduOcrException(
        tokenResponse.data?['error_description']?.toString() ?? '百度 OCR 鉴权失败',
      );
    }

    final response = await _dio.post<Map<String, dynamic>>(
      'https://aip.baidubce.com/rest/2.0/ocr/v1/train_ticket',
      queryParameters: {'access_token': accessToken},
      data: {'image': base64Encode(imageBytes)},
      options: Options(contentType: Headers.formUrlEncodedContentType),
    );
    final data = response.data ?? const <String, dynamic>{};
    if (data['error_code'] != null) {
      throw BaiduOcrException(data['error_msg']?.toString() ?? '火车票识别失败');
    }
    final words = data['words_result'];
    if (words is! Map) throw const BaiduOcrException('未识别到火车票信息');
    final values = Map<String, dynamic>.from(words);
    final trainNumber = _word(values, const [
      'train_num',
      'train_number',
    ]).replaceAll(RegExp(r'\s|次$'), '').toUpperCase();
    final fromStation = _word(values, const [
      'starting_station',
      'start_station',
    ]);
    final toStation = _word(values, const [
      'destination_station',
      'end_station',
    ]);
    final date = _word(values, const ['date', 'ticket_date']);
    if (trainNumber.isEmpty ||
        fromStation.isEmpty ||
        toStation.isEmpty ||
        date.isEmpty) {
      throw const BaiduOcrException('车次、始发站、终到站或乘车日期识别不完整，请重新拍摄');
    }
    return TrainTicketOcrResult(
      trainNumber: trainNumber,
      fromStation: fromStation,
      toStation: toStation,
      departureTime: _parseDateTime(date),
      seatType: _normalizeSeatType(
        _nullableWord(values, const ['seat_category', 'seat_type']),
      ),
      seatNumber: _nullableWord(values, const ['seat_num', 'seat_number']),
      price: _parsePrice(
        _word(values, const ['ticket_rates', 'ticket_price', 'price']),
      ),
    );
  }

  static String _word(Map<String, dynamic> words, List<String> keys) {
    for (final key in keys) {
      final value = words[key];
      if (value is Map) {
        final text = value['words']?.toString().trim() ?? '';
        if (text.isNotEmpty) return text;
      } else {
        final text = value?.toString().trim() ?? '';
        if (text.isNotEmpty) return text;
      }
    }
    return '';
  }

  static String? _nullableWord(Map<String, dynamic> words, List<String> keys) {
    final value = _word(words, keys);
    return value.isEmpty ? null : value;
  }

  static String? _normalizeSeatType(String? value) {
    final normalized = value?.replaceAll('新空调', '').trim() ?? '';
    return normalized.isEmpty ? null : normalized;
  }

  static DateTime _parseDateTime(String value) {
    final match = RegExp(
      r'(\d{4})\D+(\d{1,2})\D+(\d{1,2})(?:\D+(\d{1,2})[:：](\d{1,2}))?',
    ).firstMatch(value);
    if (match == null) throw const BaiduOcrException('无法解析乘车日期');
    return DateTime(
      int.parse(match.group(1)!),
      int.parse(match.group(2)!),
      int.parse(match.group(3)!),
      int.tryParse(match.group(4) ?? '') ?? 0,
      int.tryParse(match.group(5) ?? '') ?? 0,
    );
  }

  static double? _parsePrice(String value) {
    return double.tryParse(value.replaceAll(RegExp(r'[^\d.]'), ''));
  }
}

class BaiduOcrException implements Exception {
  const BaiduOcrException(this.message);
  final String message;
  @override
  String toString() => message;
}
